# ¡Hola, Jorge!  

Mi nombre es Carlos Ortiz, soy code reviewer de TripleTen y voy a revisar el proyecto que acabas de desarrollar.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión. 

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>
¡Empecemos!

# PROYECTO SPRINT 9: APRENDIZAJE AUTOMATICO EN NEGOCIOS

## Descargar y preparar los datos.

In [1]:
# Inicialización 
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error


In [2]:
# Cargar los archivos de datos
df1 = pd.read_csv('/datasets/geo_data_0.csv')
df2 = pd.read_csv('/datasets/geo_data_1.csv')
df3 = pd.read_csv('/datasets/geo_data_2.csv')

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con la importación de datos y de librerías.
</div>

In [3]:
# Imprimir la información general/resumida sobre df1
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
# Imprimir una muestra de los datos de df1
df1.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [5]:
# Imprimir la información general/resumida sobre df2
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [6]:
# Imprimir una muestra de los datos de df2
df2.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [7]:
# Imprimir la información general/resumida sobre df1
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [8]:
# Imprimir una muestra de los datos de df3
df3.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Se descargaron los 3 archivos geo_data_0.csv, geo_data_1.csv, geo_data_2.csv con los datos de exploración geologica de las 3 regiones y se almacenaron en la variables df1, df2, df3 correspondientemente. Se mostraron los 5 primeros elementos de cada tabla, observando que los 3 conjuntos de datos tienen 10,000 entradas y se dividen en 5 columnas cada uno.

In [9]:
#Verificar valores nulos y duplicados
total_nulos1 = df1.isnull().sum().sum()
print("Total de valores nulos en df1:", total_nulos1)
total_duplicados1 = df1.duplicated().any().sum()
print("Total de filas duplicadas en df1:", total_duplicados1)
total_nulos2 = df2.isnull().sum().sum()
print("\nTotal de valores nulos en df2:", total_nulos2)
total_duplicados2 = df2.duplicated().any().sum()
print("Total de filas duplicadas en df2:", total_duplicados2)
total_nulos3 = df3.isnull().sum().sum()
print("\nTotal de valores nulos en df3:", total_nulos3)
total_duplicados3 = df3.duplicated().any().sum()
print("Total de filas duplicadas en df3:", total_duplicados3)

Total de valores nulos en df1: 0
Total de filas duplicadas en df1: 0

Total de valores nulos en df2: 0
Total de filas duplicadas en df2: 0

Total de valores nulos en df3: 0
Total de filas duplicadas en df3: 0


In [10]:
#Eliminar columnas innecesarias de df1
columnas_a_eliminar = ['id']
region_1 = df1.drop(columnas_a_eliminar, axis=1)
region_1.head()

f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647

In [11]:
#Eliminar columnas innecesarias de df2
columnas_a_eliminar = ['id']
region_2 = df2.drop(columnas_a_eliminar, axis=1)
region_2.head()

f0         f1        f2     product
0 -15.001348  -8.276000 -0.005876    3.179103
1  14.272088  -3.475083  0.999183   26.953261
2   6.263187  -5.948386  5.001160  134.766305
3 -13.081196 -11.506057  4.999415  137.945408
4  12.702195  -8.147433  5.004363  134.766305

In [12]:
#Eliminar columnas innecesarias de df3
columnas_a_eliminar = ['id']
region_3 = df3.drop(columnas_a_eliminar, axis=1)
region_3.head()

f0        f1        f2     product
0 -1.146987  0.963328 -0.828965   27.758673
1  0.262778  0.269839 -2.530187   56.069697
2  0.194587  0.289035 -5.586433   62.871910
3  2.236060 -0.553760  0.930038  114.572842
4 -0.515993  1.716266  5.899011  149.600746

Se verifico que en los 3 conjuntos de datos no tuvieran valores nulos ni valores duplicados. Tambien se elimino en los 3 conjuntos de datos la columna de 'id' para un mejor analisis de los datos.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Has hecho bien con esta exploración inicial de la información.
</div>

## Entrenar y probar el modelo en cada región

### Dividir los datos en conjunto de entrenamiento y  conjunto de validación en proporcion de 75:25 

In [13]:
#Segmentar los datos en la región 1
target1 = region_1['product']
features1 = region_1.drop(['product'], axis=1)
features_train1, features_valid1, target_train1, target_valid1 = train_test_split(features1, target1, test_size=0.25, random_state=88)
    
print('Dataset de entrenamiento de la región 1', features_train1.shape, target_train1.shape)
print('Dataset de validacion de la región 1', features_valid1.shape, target_valid1.shape)

Dataset de entrenamiento de la región 1 (75000, 3) (75000,)
Dataset de validacion de la región 1 (25000, 3) (25000,)


In [14]:
#Segmentar los datos en la región 2
target2 = region_2['product']
features2 = region_2.drop(['product'], axis=1)
features_train2, features_valid2, target_train2, target_valid2 = train_test_split(features2, target2, test_size=0.25, random_state=88)
    
print('Dataset de entrenamiento de la región 2', features_train2.shape, target_train2.shape)
print('Dataset de validacion de la región 2', features_valid2.shape, target_valid2.shape)

Dataset de entrenamiento de la región 2 (75000, 3) (75000,)
Dataset de validacion de la región 2 (25000, 3) (25000,)


In [15]:
#Segmentar los datos en la región 3
target3 = region_3['product']
features3 = region_3.drop(['product'], axis=1)
features_train3, features_valid3, target_train3, target_valid3 = train_test_split(features3, target3, test_size=0.25, random_state=88)
    
print('Dataset de entrenamiento de la región 3', features_train3.shape, target_train3.shape)
print('Dataset de validacion de la región 3', features_valid3.shape, target_valid3.shape)

Dataset de entrenamiento de la región 3 (75000, 3) (75000,)
Dataset de validacion de la región 3 (25000, 3) (25000,)


Se dividieron los 3 conjuntos de datos en 2 partes: en dataset de entrenamiento (75% del total de datos) y dataset de validación (25% del total de datos).

### Entrenar el modelo y hacer predicciones para el conjunto de validación

In [16]:
#Entrenar el modelo y hacer predicciones en las región 1
model_region1 = LinearRegression()
model_region1.fit(features_train1, target_train1)
predicted_valid1 = model_region1.predict(features_valid1)


In [17]:
#Entrenar el modelo y hacer predicciones en las región 2
model_region2 = LinearRegression()
model_region2.fit(features_train2, target_train2)
predicted_valid2 = model_region2.predict(features_valid2)


In [18]:
#Entrenar el modelo y hacer predicciones en las región 1
model_region3 = LinearRegression()
model_region3.fit(features_train3, target_train3)
predicted_valid3 = model_region3.predict(features_valid3)


Entrenamos los 3 conjuntos de datos en el dataset de entrenamiento utilizando la regresion lineal ya que el objetivo es númerico 

### Guardar las predicciones y respuestas correctas para el conjunto de validación

In [19]:
#Guardar predicciones y respuestas correctas en la región 1
resultados1 = pd.DataFrame({'predicciones': predicted_valid1, 'respuestas_correctas': target_valid1.values})
resultados1.head()

predicciones  respuestas_correctas
0    121.234306            123.801417
1     79.036922             30.132880
2    114.655055            157.733201
3     97.422219            140.020959
4     85.743222            104.368564

In [20]:
#Guardar predicciones y respuestas correctas en la región 2
resultados2 = pd.DataFrame({'predicciones': predicted_valid2, 'respuestas_correctas': target_valid2.values})
resultados2.head()

predicciones  respuestas_correctas
0    134.662513            134.766305
1    108.418348            107.813044
2     56.612666             57.085625
3     81.537792             80.859783
4     81.124996             80.859783

In [21]:
#Guardar predicciones y respuestas correctas en la región 3
resultados3 = pd.DataFrame({'predicciones': predicted_valid3, 'respuestas_correctas': target_valid3.values})
resultados3.head()

predicciones  respuestas_correctas
0    129.396185            168.989257
1    120.201671             89.151554
2    115.891075            141.126068
3     82.396435            156.962126
4     90.837385            106.932658

Guardamos los resultados de las 3 regiones en un DataFrame cada uno, con las predicciones y respuestas correctas para el conjunto de validación.  

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen trabajo en esta sección.
</div>

### Mostrar el volumen medio de reservas predicho y RMSE del modelo

In [22]:
#Volumen medio de reservas predicho y RMSE del modelo de la región 1
volumen_medio_reservas1 = resultados1['predicciones'].mean()
print("Volumen medio de reservas de la región 1 =", volumen_medio_reservas1)

mse1 = mean_squared_error(target_valid1, predicted_valid1)
print("Raíz Cuadrada del Error Cuadratico Medio (RMSE) de la región 1 =", mse1 ** 0.5)

Volumen medio de reservas de la región 1 = 92.61602155606631
Raíz Cuadrada del Error Cuadratico Medio (RMSE) de la región 1 = 37.37693856838324


In [23]:
#Volumen medio de reservas predicho y RMSE del modelo de la región 2
volumen_medio_reservas2 = resultados2['predicciones'].mean()
print("Volumen medio de reservas de la región 2 =", volumen_medio_reservas2)

mse2 = mean_squared_error(target_valid2, predicted_valid2)
print("Raíz Cuadrada del Error Cuadratico Medio (RMSE) de la región 2 =", mse2 ** 0.5)

Volumen medio de reservas de la región 2 = 68.8413655286577
Raíz Cuadrada del Error Cuadratico Medio (RMSE) de la región 2 = 0.8976095608958887


In [24]:
#Volumen medio de reservas predicho y RMSE del modelo de la región 3
volumen_medio_reservas3 = resultados3['predicciones'].mean()
print("Volumen medio de reservas de la región 3 =", volumen_medio_reservas3)

mse3 = mean_squared_error(target_valid3, predicted_valid3)
print("Raíz Cuadrada del Error Cuadratico Medio (RMSE) de la región 3 =", mse3 ** 0.5)

Volumen medio de reservas de la región 3 = 95.18516847231068
Raíz Cuadrada del Error Cuadratico Medio (RMSE) de la región 3 = 40.15761558006934


Calculamos para cada región el MSE (Mean Squared Error) y RMSE (Root Mean Squared Error) que son métricas comunes utilizadas para evaluar el rendimiento de modelos de regresión

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
¡Perfecto!
</div>

### Analizar los resultados

Calculando el RMSE para las 3 regiones encontramos que el valor mas cercano a 0 es en la región 2, con un valor RMSE = 0.8976095608958887, esto indica un mejor ajuste del modelo a los datos, en cambio las regiones 1 y 3 estan mas alejadas del valor 0 indicando un mayor error entre las predicciones y los valores reales, lo que implica un peor ajuste del modelo

### Colocar pasos previos en funciones y ejecutarlos en todas las regiones

In [25]:
#Función para todas las regiones
def entrenamiento_modelo_regiones(data):
    
    target = data['product']
    features = data.drop(['product'], axis=1)
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=88)
    
    model = LinearRegression()
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    
    resultados = pd.DataFrame({'predicciones': predicted_valid, 'respuestas_correctas': target_valid.values})
    
    volumen_medio_reservas = resultados['predicciones'].mean()
    mse = mean_squared_error(target_valid, predicted_valid)
    rmse = mse ** 0.5
    
    print("Volumen medio de reservas =", volumen_medio_reservas)
    print("Raíz Cuadrada del Error Cuadratico Medio (RMSE) =", rmse)
    
    return resultados 
    
resultados_region_1 = entrenamiento_modelo_regiones(region_1)
resultados_region_2 = entrenamiento_modelo_regiones(region_2)
resultados_region_3 = entrenamiento_modelo_regiones(region_3)    

Volumen medio de reservas = 92.61602155606631
Raíz Cuadrada del Error Cuadratico Medio (RMSE) = 37.37693856838324
Volumen medio de reservas = 68.8413655286577
Raíz Cuadrada del Error Cuadratico Medio (RMSE) = 0.8976095608958887
Volumen medio de reservas = 95.18516847231068
Raíz Cuadrada del Error Cuadratico Medio (RMSE) = 40.15761558006934


Creamos la función entrenamiento_modelo_regiones  para calcular el MSE y RMSE de las 3 regiones en un solo paso para simplificar nuestro codigo.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Bien. Siempre es mejor aprovechas las oportunidades de modularización.
</div>

## Calculo de ganancias

### Almacenar todos los valores necesarios para los calculos en variables separadas

In [26]:
#Almacenar valores en variables
volumen_medio_reservas_region_1 = 92.61602155606631
volumen_medio_reservas_region_2 = 68.8413655286577
volumen_medio_reservas_region_3 = 95.18516847231068


Almacenamos los valores necesarios de acuerdo a las condiciones del proyecto para facilitar calculos posteriores

### Comparar cantidades 

In [27]:
#Condiciones del negocio
presupuesto = 100_000_000
numero_de_pozos = 200
unidad_barril = 4_500
unidades_evitar_perdidas = 111.1

In [28]:
#Comparar cantidades (unidades_evitar_perdidas con cantidad media de reservas en cada región)
unidades_region_1 = unidades_evitar_perdidas - volumen_medio_reservas_region_1
unidades_region_2 = unidades_evitar_perdidas - volumen_medio_reservas_region_2
unidades_region_3 = unidades_evitar_perdidas - volumen_medio_reservas_region_3

print('Región 1 =', unidades_region_1, 'unidades faltantes')
print('Región 2 =', unidades_region_2, 'unidades faltantes')
print('Región 3 =', unidades_region_3, 'unidades faltantes')




Región 1 = 18.48397844393368 unidades faltantes
Región 2 = 42.25863447134229 unidades faltantes
Región 3 = 15.914831527689316 unidades faltantes


Comparamos las cantidades de las 3 regiones observando que las 1 y 3 son las que menos unidades faltantes requieren para tener perdidas.

### Presentar conclusiones sobre como preparar el paso para calcular el beneficio

De acuerdo con la comparacion de datos en las 3 regiones, observamos que en ninguna de las regiones tenemos las suficientes unidades requeridas para tener ganancias, para preparar el beneficio buscaremos los 200 pozos con los valores de predicción mas altos.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

## Escribir una función para calcular la ganancia en cada región

### Elegir los 200 pozos con los valores de predicción mas altos 

In [29]:
#Ordenar las predicciones en la region 1
top_200_pozos_r1 = resultados1.sort_values(by='predicciones', ascending=False)
top_200_pozos_r1 = top_200_pozos_r1.iloc[:200]
top_200_pozos_r1



predicciones  respuestas_correctas
10250    177.534575            123.920559
15417    176.072505            140.631646
18321    175.919680            162.153488
831      173.913571            145.106893
8050     172.689685            176.807828
...             ...                   ...
7619     147.730639            162.897833
23303    147.702197            120.776707
9370     147.679288            138.886443
18267    147.666609            148.408951
21891    147.644635            150.835439

[200 rows x 2 columns]

In [30]:
#Ordenar las predicciones en la region 2
top_200_pozos_r2 = resultados2.sort_values(by='predicciones', ascending=False)
top_200_pozos_r2 = top_200_pozos_r2.iloc[:200]
top_200_pozos_r2

predicciones  respuestas_correctas
9635     139.895597            137.945408
6753     139.662139            137.945408
18149    139.550920            137.945408
15615    139.466759            137.945408
14508    139.379047            137.945408
...             ...                   ...
17881    138.368938            137.945408
15337    138.368859            137.945408
9475     138.368084            137.945408
20752    138.362241            137.945408
8559     138.362094            137.945408

[200 rows x 2 columns]

In [31]:
#Ordenar las predicciones en la region 1
top_200_pozos_r3 = resultados3.sort_values(by='predicciones', ascending=False)
top_200_pozos_r3 = top_200_pozos_r3.iloc[:200]
top_200_pozos_r3

predicciones  respuestas_correctas
9812     173.995419            152.294162
13939    171.590606            139.562959
11867    166.007074            131.627481
9697     164.582479             97.775979
13811    164.061527            125.535602
...             ...                   ...
5528     143.007042            140.560911
16087    142.987208            165.714444
9214     142.909578            177.344843
13763    142.846196            136.914899
16455    142.812437             89.492500

[200 rows x 2 columns]

Para obtener los 200 pozos con los valores de predicción mas altos, ordenamos en forma descendente los valores de la columna 'predicciones' mas altos y los guardamos en las variables de top_200_pozos en cada una de las 3 regiones.

### Resumir el volumen objetivo de reservas de dichas predicciones

In [32]:
#Encontrar el volumen objetivo en la region 1 
respuestas_correctas_r1 = top_200_pozos_r1['respuestas_correctas']
respuestas_correctas_r1.sum()
volumen_region1 = respuestas_correctas_r1 * unidad_barril  
print('Volumen objetivo de reservas de dichas predicciones en la región 1 =', volumen_region1.sum())

Volumen objetivo de reservas de dichas predicciones en la región 1 = 133135684.39095855


In [33]:
#Encontrar el volumen objetivo en la region 2
respuestas_correctas_r2 = top_200_pozos_r2['respuestas_correctas']
respuestas_correctas_r2.sum()
volumen_region2 = respuestas_correctas_r2 * unidad_barril  
print('Volumen objetivo de reservas de dichas predicciones en la región 2 =', volumen_region2.sum())

Volumen objetivo de reservas de dichas predicciones en la región 2 = 124150866.96681508


In [34]:
#Encontrar el volumen objetivo en la region 3
respuestas_correctas_r3 = top_200_pozos_r3['respuestas_correctas']
respuestas_correctas_r3.sum()
volumen_region3 = respuestas_correctas_r3 * unidad_barril  
print('Volumen objetivo de reservas de dichas predicciones en la región 3 =', volumen_region3.sum())

Volumen objetivo de reservas de dichas predicciones en la región 3 = 125311974.97581083


Encontramos el volumen objetivo de reservas de las predicciones en cada región sumando el total de los 200 pozos mas altos y multiplicando la suma por unidad de barril, siendo la region 1 con el valor mas alto de 133135684.39095855

### Calcular la ganancia potencial de los 200 pozos principales

In [35]:
#Ganancia potencial en la region 1
ganancia_r1 = volumen_region1.sum() - presupuesto
print('Ganancia potencial de los 200 pozos principales en la región 1 =', ganancia_r1)

Ganancia potencial de los 200 pozos principales en la región 1 = 33135684.390958548


In [36]:
#Ganancia potencial en la region 2
ganancia_r2 = volumen_region2.sum() - presupuesto
print('Ganancia potencial de los 200 pozos principales en la región 2 =', ganancia_r2)

Ganancia potencial de los 200 pozos principales en la región 2 = 24150866.966815084


In [37]:
#Ganancia potencial en la region 3
ganancia_r3 = volumen_region3.sum() - presupuesto
print('Ganancia potencial de los 200 pozos principales en la región 3 =', ganancia_r3)

Ganancia potencial de los 200 pozos principales en la región 3 = 25311974.975810826


Calculamos la ganancia potencial en cada region, obteniendo la región 1 con mayor ganancia con 33135684.390958548, le sigue la región 3 con 25311974.975810826, y al ultimo la región 2 con 24150866.966815084

### Función para calcular ganancia portencial en las 3 regiones

In [38]:
#Funcion ganancia
def ganancia_regiones(data):
    
    top_200_pozos = data.sort_values(by='predicciones', ascending=False)
    top_200_pozos = top_200_pozos.iloc[:200]
    
    respuestas_correctas = top_200_pozos['respuestas_correctas']
    respuestas_correctas.sum()
    volumen_region = respuestas_correctas.sum() * unidad_barril
    
    ganancia = volumen_region - presupuesto
    
    print("Ganancia region =", ganancia)
    
    return ganancia

ganancia_region_1 = ganancia_regiones(resultados_region_1)
ganancia_region_2 = ganancia_regiones(resultados_region_2)
ganancia_region_3 = ganancia_regiones(resultados_region_3)

Ganancia region = 33135684.390958548
Ganancia region = 24150866.966815114
Ganancia region = 25311974.975810826


Creamos la funcion ganancia_regiones para calcular la ganancia en las 3 regiones en un solo paso, para simplificar el codigo.

## Calcula Riesgos y ganancias de cada región

### Emplear la tecnica de Bootstrapping con 1000 muestras para hallar la distribución de los beneficios

In [39]:
#Aplicar tencica Bootstrapping en la region 1
beneficios_region1 =[]

for i in range(1000):
    muestra_region1 = resultados1.sample(n=500, replace=True)
    beneficios_region1.append(muestra_region1)
    

In [40]:
#Aplicar tencica Bootstrapping en la region 2
beneficios_region2 =[]

for i in range(1000):
    muestra_region2 = resultados2.sample(n=500, replace=True)
    beneficios_region2.append(muestra_region2)

In [41]:
#Aplicar tencica Bootstrapping en la region 3
beneficios_region3 =[]

for i in range(1000):
    muestra_region3 = resultados3.sample(n=500, replace=True)
    beneficios_region3.append(muestra_region3)

In [42]:
#Función para calcular el beneficio y hallar la distribución de cada region.
def calcular_beneficio(data):
    top_200_pozos = data.sort_values(by='predicciones', ascending=False)
    top_200_pozos = top_200_pozos.iloc[:200]
    
    respuestas_correctas = top_200_pozos['respuestas_correctas']
    volumen_region = respuestas_correctas.sum() * unidad_barril
    
    calculo_beneficio = volumen_region - presupuesto
    
    return calculo_beneficio

In [49]:
# Calcular beneficio para cada submuestra y almacenar la distribucion de los beneficios en la región 1
calculo_beneficios_region1 = pd.Series([calcular_beneficio(submuestra) for submuestra in beneficios_region1])
calculo_beneficios_region1


0      6.116924e+06
1      5.375016e+06
2     -6.142305e+05
3      4.390339e+06
4      2.458961e+06
           ...     
995    7.608537e+06
996    5.331855e+06
997    3.648286e+06
998    6.181332e+06
999    4.351970e+06
Length: 1000, dtype: float64

In [50]:
# Calcular beneficio para cada submuestra y almacenar la distribucion de los beneficios en la región 2
calculo_beneficios_region2 = pd.Series([calcular_beneficio(submuestra) for submuestra in beneficios_region2])
calculo_beneficios_region2

0      4.998294e+06
1      4.577201e+06
2      8.851574e+06
3      1.202861e+06
4      5.904546e+06
           ...     
995    5.440535e+06
996    3.600042e+06
997    2.522741e+06
998    1.816151e+06
999    5.739716e+06
Length: 1000, dtype: float64

In [51]:
# Calcular beneficio para cada submuestra y almacenar la distribucion de los beneficios en la región 3
calculo_beneficios_region3 = pd.Series([calcular_beneficio(submuestra) for submuestra in beneficios_region3])
calculo_beneficios_region3

0      4.728955e+06
1      6.854608e+06
2      5.784844e+06
3      3.971913e+06
4      5.858295e+06
           ...     
995    5.287002e+06
996    5.763765e+06
997    4.683416e+06
998    5.403784e+06
999    4.370819e+06
Length: 1000, dtype: float64

Se trabajo con el DataFrame de cada región que contenia las predicciones y respuestas correctas, utilizando la tecnica de Bootstrapping creamos un muestreo estadistico utilizando 500 elementos de nuestros DataFrame para formar una submuestra y calculamos el beneficio de nuestra muestra resultante, iterando 1000 veces para encontar la distribucion de los beneficios.

### Encontrar el beneficio promedio, el intervalo de confianza del 95% y el riesgo de perdidas

In [46]:
#Calcular la media de los beneficios en la región 1
media_beneficios1 = np.mean(calculo_beneficios_region1)

#Calcular los intervalos de confianza al 95% en la región 1
cuartil_2_5_1 = calculo_beneficios_region1.quantile(0.025)
cuartil_97_5_1 = calculo_beneficios_region1.quantile(0.975)

#Calcular la probabilidad de pérdida (beneficio negativo) en la region 1
probabilidad_perdida1 = np.mean(np.array(calculo_beneficios_region1) < 0)
porcentaje_perdida1 = probabilidad_perdida1 * 100

#Imprimir resultados
print("Beneficio Promedio en la Región 1 =", media_beneficios1)
print("Intervalo de Confianza al 95% en la Región 1 = (2.5)", cuartil_2_5_1, "- (97.5)", cuartil_97_5_1)
print("Riesgo de Pérdida en la Región 1:", porcentaje_perdida1, "%")

Beneficio Promedio en la Región 1 = 4146669.604236317
Intervalo de Confianza al 95% en la Región 1 = (2.5) -1021181.3308986183 - (97.5) 8947153.766743207
Riesgo de Pérdida en la Región 1: 6.0 %


In [47]:
#Calcular la media de los beneficios en la región 2
media_beneficios2 = np.mean(calculo_beneficios_region2)

#Calcular los intervalos de confianza al 95% en la región 2
cuartil_2_5_2 = calculo_beneficios_region2.quantile(0.025)
cuartil_97_5_2 = calculo_beneficios_region2.quantile(0.975)

#Calcular la probabilidad de pérdida (beneficio negativo) en la region 2
probabilidad_perdida2 = np.mean(np.array(calculo_beneficios_region2) < 0)
porcentaje_perdida2 = probabilidad_perdida2 * 100

#Imprimir resultados
print("Beneficio Promedio en la Región 2 =", media_beneficios2)
print("Intervalo de Confianza al 95% en la Región 2 = (2.5)", cuartil_2_5_2, "- (97.5)", cuartil_97_5_2)
print("Riesgo de Pérdida en la Región 2:", porcentaje_perdida2, "%")

Beneficio Promedio en la Región 2 = 4462830.285733135
Intervalo de Confianza al 95% en la Región 2 = (2.5) 260362.00280298182 - (97.5) 8380627.970288682
Riesgo de Pérdida en la Región 2: 1.5 %


In [48]:
#Calcular la media de los beneficios en la región 3
media_beneficios3 = np.mean(calculo_beneficios_region3)

#Calcular los intervalos de confianza al 95% en la región 3
cuartil_2_5_3 = calculo_beneficios_region3.quantile(0.025)
cuartil_97_5_3 = calculo_beneficios_region3.quantile(0.975)

#Calcular la probabilidad de pérdida (beneficio negativo) en la region 3
probabilidad_perdida3 = np.mean(np.array(calculo_beneficios_region3) < 0)
porcentaje_perdida3 = probabilidad_perdida3 * 100

#Imprimir resultados
print("Beneficio Promedio en la Región 3 =", media_beneficios3)
print("Intervalo de Confianza al 95% en la Región 3 = (2.5)", cuartil_2_5_3, "- (97.5)", cuartil_97_5_3)
print("Riesgo de Pérdida en la Región 3:", porcentaje_perdida3, "%")

Beneficio Promedio en la Región 3 = 3684550.817087282
Intervalo de Confianza al 95% en la Región 3 = (2.5) -1363089.6200280425 - (97.5) 8898318.34748069
Riesgo de Pérdida en la Región 3: 7.000000000000001 %


Encontramos el beneficio promedio y el resgo de perdidas con un intervalo de confianza del 95%, la perdida de ganancia la calculamos como una probabilidad y la expresamos como porcentaje.

### Conclusiones

De acuerdo al analisis que se hizo, la lista con ganancias potenciales de los 200 pozos con los valores de predicción mas altos es la siguiente:
- Ganancia region 1 = 33135684.390958548
- Ganancia region 2 = 24150866.966815114
- Ganancia region 3 = 25311974.975810826

Se hizo un análisis mas profundo de los datos en las 3 regiones con la técnica estadística de remuestreo Bootstrapping creando un muestreo estadistico utilizando 500 elementos de forma aleatoria de nuestros datos para formar 1 submuestra y a su vez iterarla en 1000 ocasiones para calcular el beneficio promedio, utilizando un intervalo de confianza del 95% y el porcentaje de riego de perdidaobteniendo la siguiente lista:
- Región 1:
- Beneficio Promedio = 4146669.604236317
- Intervalo de Confianza al 95% = (2.5) -1021181.3308986183 - (97.5) 8947153.766743207
- Riesgo de Pérdida: 6.0 %

- Región 2
- Beneficio Promedio = 4462830.285733135
- Intervalo de Confianza al 95% = (2.5) 260362.00280298182 - (97.5) 8380627.970288682
- Riesgo de Pérdida: 1.5 %

- Región 3
- Beneficio Promedio = 3684550.817087282
- Intervalo de Confianza al 95% = (2.5) -1363089.6200280425 - (97.5) 8898318.34748069
- Riesgo de Pérdida: 7.000000000000001 %

De acuerdo al analisis de datos se propone que la región 2 es la mas adecuada para el desarrollo de pozos petroliferos ya que tiene una ganancia de 4462830.285733135 que es mayor con respecto a las otra 2 regiones y el riesgo de perdida es mucho menor con tan solo el 1.5%. Tambien observamos que en el intervalo de confianza del 95%  en la parte inferior del intervalo no presenta numero negativo. 

<div class="alert alert-block alert-success">

# Comentarios generales
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Todo corregido. Has aprobado un nuevo proyecto. ¡Felicitaciones!
</div>